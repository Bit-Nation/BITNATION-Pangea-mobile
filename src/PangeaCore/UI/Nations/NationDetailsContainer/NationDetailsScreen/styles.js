import { MediaQueryStyleSheet } from 'react-native-responsive';
import GlobalStyles from 'pangea-common-reactnative/styles';

const styles = MediaQueryStyleSheet.create({
  ...GlobalStyles,

  scrollView: {
    flex: 1,
  },

});

export default styles;
