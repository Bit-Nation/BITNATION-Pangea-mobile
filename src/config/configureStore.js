// @flow

import { createStore, applyMiddleware, compose, type Store } from 'redux';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import createSagaMonitor from '@clarketm/saga-monitor';
import reducers from '../reducers';
import rootSaga from '../sagas';

/**
 * @desc Configures a Redux store.
 * @return {Store} Created store object.
 */
export default function configureStore(): Store {
  const sagaMonitor = createSagaMonitor({
    level: 'log',
    actionDispatch: true,
  });
  const sagaMiddleware = createSagaMiddleware({ sagaMonitor });

  const enhancer = compose(
    applyMiddleware(sagaMiddleware),
    applyMiddleware(logger),
  );

  const store = createStore(reducers, enhancer);
  sagaMiddleware.run(rootSaga);
  return store;
}
