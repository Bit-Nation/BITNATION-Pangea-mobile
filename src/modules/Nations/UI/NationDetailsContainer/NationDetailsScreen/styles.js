import { MediaQueryStyleSheet } from 'react-native-responsive';
import GlobalStyles from 'pangea-common-reactnative/Styles';

const styles = MediaQueryStyleSheet.create({
  ...GlobalStyles,

  scrollView: {
    flex: 1,
  },

});

export default styles;
