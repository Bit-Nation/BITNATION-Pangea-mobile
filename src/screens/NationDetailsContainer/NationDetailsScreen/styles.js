import { MediaQueryStyleSheet } from 'react-native-responsive'
import Colors from '../../../global/Colors'
import GlobalStyles from '../../../global/Styles'

const styles = MediaQueryStyleSheet.create({
	...GlobalStyles,
	
	title: {
		...GlobalStyles.largeTitle,
	},
	
	scrollView: {
		flex: 1,
	},
	
	messageView: {
		marginBottom: 8,
		marginLeft: 8,
		marginRight: 8,
	},
	
})

export default styles
