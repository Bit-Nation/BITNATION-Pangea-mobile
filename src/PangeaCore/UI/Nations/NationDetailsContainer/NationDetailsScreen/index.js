// @flow

import React, { Component } from 'react';
import {
  View,
  Text, ScrollView,
} from 'react-native';

import BackgroundImage from 'pangea-common-reactnative/UI/BackgroundImage';
import styles from './styles';
import NationActionButton from 'pangea-common-reactnative/UI/NationActionButton';
import AssetsImage from 'pangea-common-reactnative/assets/AssetsImages';
import PanelView from 'pangea-common-reactnative/UI/PanelView';
import FakeNavigationBar from 'pangea-common-reactnative/UI/FakeNavigationBar';
import i18n from 'pangea-common/i18n';
import { openedNation, type State as NationState } from '@pangea/nations/nations-reducers';
import PanelViewAlert from 'pangea-common-reactnative/UI/PanelViewAlert';
import PanelViewCitizen from 'pangea-common-reactnative/UI/PanelViewCitizen';
import { nationIsValid, resolveStatus } from '@pangea/nations/nations-utils';
import type { Navigator } from 'pangea-common-reactnative/ReactNativeNavigation-types';
import type { NationType } from '@pangea/nations/Nation-types';
import ScreenTitle from 'pangea-common-reactnative/UI/ScreenTitle';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc A boolean prop to indicate whether the nation is Draft or not
   */
  isDraft: boolean,
};

type Actions = {
  /**
   * @desc Function to join a nation
   */
  joinNation: () => void,
  /**
   * @desc Function to leave a nation
   */
  leaveNation: () => void,
  /**
   * @desc Function to delete the draft of a nation
   */
  deleteDraft: () => void,
  /**
   * @desc Function to submit a nation to the blockchain
   */
  submitDraft: () => void,
  /**
   * @desc Function to open a nation's chat
   */
  openNationChat: () => void,
}

class NationDetailsScreen extends Component<Props & Actions & NationState> {
  static disableJoinButton(nation: NationType) {
    if (nation.tx && nation.tx.status === 200) {
      return true;
    }

    if (nation.joined === true) {
      return true;
    }

    return false;
  }

  static disableLeaveButton(nation: NationType) {
    if (nation.tx && nation.tx.status === 200) {
      return true;
    }

    if (nation.joined === false) {
      return true;
    }

    return false;
  }

  static buildAboutView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('screens.nationDetails.aboutInfo', { name: nation.nationName })}
      >
        <Text style={styles.panelSubTitle}>
          {`${i18n.t('screens.nationDetails.description')}:`}
        </Text>
        <Text style={styles.body}>
          {nation.nationDescription ? `${nation.nationDescription}\n` : ''}
          {`${i18n.t('screens.nationDetails.locationInfo', {
            name: nation.nationName,
            locationType: nation.exists ?
              i18n.t('enums.nation.locationType.geographical') :
              i18n.t('enums.nation.locationType.virtual'),
          })}\n`}
        </Text>
        <Text style={styles.panelSubTitle}>
          {i18n.t('screens.nationDetails.ethereumAddress')}
        </Text>
        <Text style={styles.footnote}>
          {nation.ethAddress}
        </Text>
      </PanelView>
    );
  }

  static buildGovernmentalStructureView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('common.governmentalStructure')}
      >
        <Text style={styles.body}>
          {i18n.t('screens.nationDetails.legalSystemInfo', {
            name: nation.nationName,
            code: nation.nationCode,
          })}
          {' '}
          {i18n.t('screens.nationDetails.lawEnforcementInfo', {
            lawEnforcementMechanism: nation.lawEnforcementMechanism,
          })}
          {' '}
          {i18n.t('screens.nationDetails.governmentInfo', {
            decisionMakingProcess: nation.decisionMakingProcess,
          })}
        </Text>
      </PanelView>
    );
  }

  static buildFactsView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('screens.nationDetails.funFacts')}
      >
        <Text style={styles.body}>
          {nation.diplomaticRecognition ? (`${i18n.t('screens.nationDetails.diplomaticRecognitionInfo', { name: nation.nationName })}\n\n`) : ''}

          {i18n.t('screens.nationDetails.serviceUsageInfo', {
            allowance: nation.nonCitizenUse ? i18n.t('enums.nation.usageAllowance.may') : i18n.t('enums.nation.usageAllowance.mayNot'),
          })}
          {'\n\n'}

          {i18n.t('screens.nationDetails.profitInfo', {
            profit: nation.profit ? i18n.t('enums.nation.profit.for') : i18n.t('enums.nation.profit.non'),
          })}
        </Text>
      </PanelView>
    );
  }

  static buildStatusPanel(status) {
    return (
      <PanelViewAlert
        status={status}
      />
    );
  }

  static buildCitizenPanel(nation: NationType) {
    if (nation.joined) {
      return (
        <PanelViewCitizen
          nationName={nation.nationName}
        />
      );
    }
    return null;
  }
  static defaultProps: Object;
  buildTabBar() {
    const nation = openedNation(this.props);
    if (nation === null) return null;
    if (this.props.isDraft) {
      return (
        <View style={styles.fakeBottomBar}>

          <NationActionButton
            iconSource={AssetsImage.Actions.delete}
            title={i18n.t('screens.nations.toolbar.delete')}
            disable={false}
            onPress={this.props.deleteDraft}
          />
          <NationActionButton
            iconSource={AssetsImage.Actions.submit}
            title={i18n.t('screens.nations.toolbar.submit')}
            disable={!nationIsValid(nation)}
            onPress={this.props.submitDraft}
          />
        </View>
      );
    }

    return (
      <View style={styles.fakeBottomBar}>
        <NationActionButton
          iconSource={AssetsImage.Actions.chat}
          title={i18n.t('screens.nations.toolbar.chat')}
          disable={false}
          onPress={this.props.openNationChat}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.map}
          title={i18n.t('screens.nations.toolbar.map')}
          disable
          onPress={() => {}}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.join}
          title={i18n.t('screens.nations.toolbar.join')}
          disable={NationDetailsScreen.disableJoinButton(nation)}
          onPress={this.props.joinNation}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.leave}
          title={i18n.t('screens.nations.toolbar.leave')}
          disable={NationDetailsScreen.disableLeaveButton(nation)}
          onPress={this.props.leaveNation}
        />
      </View>
    );
  }
  render() {
    const nation = openedNation(this.props);

    if (!nation) {
      this.props.navigator.pop();
      return <BackgroundImage />;
    }

    const status = resolveStatus(nation);
    const statusDescription = (status !== null ? i18n.ifExists(`screens.nationDetails.statusDescription.${status.key}`) : '');

    return (
      <View style={styles.screenContainer}>
        <BackgroundImage />
        <FakeNavigationBar />
        <View style={styles.bodyContainer}>
          <ScreenTitle title={nation.nationName} />

          <ScrollView>
            {statusDescription !== '' && NationDetailsScreen.buildStatusPanel(statusDescription)}

            {NationDetailsScreen.buildAboutView(nation)}
            {NationDetailsScreen.buildCitizenPanel(nation)}
            {NationDetailsScreen.buildGovernmentalStructureView(nation)}
            {NationDetailsScreen.buildFactsView(nation)}
          </ScrollView>
        </View>
        {this.buildTabBar()}
      </View>
    );
  }
}

NationDetailsScreen.defaultProps = {
  joinNation: () => null,
  leaveNation: () => null,
  deleteDraft: () => null,
  submitDraft: () => null,
  openNationChat: () => null,
};

export default NationDetailsScreen;
