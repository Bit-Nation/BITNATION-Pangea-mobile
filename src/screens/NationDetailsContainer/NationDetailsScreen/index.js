// @flow

import React, { Component } from 'react';
import {
  View,
  Text, ScrollView,
} from 'react-native';

import BackgroundImage from '../../../components/common/BackgroundImage';
import styles from './styles';
import NationActionButton from '../../../components/common/NationActionButton';
import AssetsImage from '../../../global/AssetsImages';
import PanelView from '../../../components/common/PanelView';
import FakeNavigationBar from '../../../components/common/FakeNavigationBar';
import i18n from '../../../global/i18n';
import { openedNation, type State as NationState } from '../../../reducers/nations';
import PanelViewAlert from '../../../components/common/PanelViewAlert';
import PanelViewCitizen from '../../../components/common/PanelViewCitizen';
import { nationIsValid, resolveStatus } from '../../../utils/nations';
import type { Navigator } from '../../../types/ReactNativeNavigation';
import type { NationType } from '../../../types/Nation';
import ScreenTitle from '../../../components/common/ScreenTitle';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc A boolean prop to indicate whether the nation is Draft or not
   */
  isDraft: boolean,
};

type Actions = {
  /**
   * @desc Function to join a nation
   */
  joinNation: () => void,
  /**
   * @desc Function to leave a nation
   */
  leaveNation: () => void,
  /**
   * @desc Function to delete the draft of a nation
   */
  deleteDraft: () => void,
  /**
   * @desc Function to submit a nation to the blockchain
   */
  submitDraft: () => void,
  /**
   * @desc Function to open a nation's chat
   */
  openNationChat: () => void,
}

class NationDetailsScreen extends Component<Props & Actions & NationState> {
  static disableJoinButton(nation: NationType) {
    if (nation.tx && nation.tx.status === 200) {
      return true;
    }

    if (nation.joined === true) {
      return true;
    }

    return false;
  }

  static disableLeaveButton(nation: NationType) {
    if (nation.tx && nation.tx.status === 200) {
      return true;
    }

    if (nation.joined === false) {
      return true;
    }

    return false;
  }

  static buildAboutView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('screens.nationDetails.aboutInfo', { name: nation.nationName })}
      >
        <Text style={styles.panelSubTitle}>
          {`${i18n.t('screens.nationDetails.description')}:`}
        </Text>
        <Text style={styles.body}>
          {nation.nationDescription ? `${nation.nationDescription}\n` : ''}
          {`${i18n.t('screens.nationDetails.locationInfo', {
            name: nation.nationName,
            locationType: nation.exists ?
              i18n.t('enums.nation.locationType.geographical') :
              i18n.t('enums.nation.locationType.virtual'),
          })}\n`}
        </Text>
        <Text style={styles.panelSubTitle}>
          {i18n.t('screens.nationDetails.ethereumAddress')}
        </Text>
        <Text style={styles.footnote}>
          {nation.ethAddress}
        </Text>
      </PanelView>
    );
  }

  static buildGovernmentalStructureView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('common.governmentalStructure')}
      >
        <Text style={styles.body}>
          {i18n.t('screens.nationDetails.legalSystemInfo', {
            name: nation.nationName,
            code: nation.nationCode,
          })}
          {' '}
          {i18n.t('screens.nationDetails.lawEnforcementInfo', {
            lawEnforcementMechanism: nation.lawEnforcementMechanism,
          })}
          {' '}
          {i18n.t('screens.nationDetails.governmentInfo', {
            decisionMakingProcess: nation.decisionMakingProcess,
          })}
        </Text>
      </PanelView>
    );
  }

  static buildFactsView(nation: NationType) {
    return (
      <PanelView
        style={styles.panelView}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('screens.nationDetails.funFacts')}
      >
        <Text style={styles.body}>
          {nation.diplomaticRecognition ? (`${i18n.t('screens.nationDetails.diplomaticRecognitionInfo', { name: nation.nationName })}\n\n`) : ''}

          {i18n.t('screens.nationDetails.serviceUsageInfo', {
            allowance: nation.createNationsNonCitizensMayUseGovernanceServices ? i18n.t('enums.nation.usageAllowance.may') : i18n.t('enums.nation.usageAllowance.mayNot'),
          })}
          {'\n\n'}

          {i18n.t('screens.nationDetails.profitInfo', {
            profit: nation.profit ? i18n.t('enums.nation.profit.for') : i18n.t('enums.nation.profit.non'),
          })}
        </Text>
      </PanelView>
    );
  }

  static buildStatusPanel(status) {
    return (
      <PanelViewAlert
        status={status}
      />
    );
  }

  static buildCitizenPanel(nation: NationType) {
    if (nation.joined) {
      return (
        <PanelViewCitizen
          nationName={nation.nationName}
        />
      );
    }
    return null;
  }
  static defaultProps: Object;
  buildTabBar() {
    const nation = openedNation(this.props);
    if (nation === null) return null;
    if (this.props.isDraft) {
      return (
        <View style={styles.fakeBottomBar}>

          <NationActionButton
            iconSource={AssetsImage.Actions.delete}
            title={i18n.t('screens.nations.toolbar.delete')}
            disable={false}
            onPress={this.props.deleteDraft}
          />
          <NationActionButton
            iconSource={AssetsImage.Actions.submit}
            title={i18n.t('screens.nations.toolbar.submit')}
            disable={!nationIsValid(nation)}
            onPress={this.props.submitDraft}
          />
        </View>
      );
    }

    return (
      <View style={styles.fakeBottomBar}>
        <NationActionButton
          iconSource={AssetsImage.Actions.chat}
          title={i18n.t('screens.nations.toolbar.chat')}
          disable={false}
          onPress={this.props.openNationChat}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.map}
          title={i18n.t('screens.nations.toolbar.map')}
          disable
          onPress={() => {}}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.join}
          title={i18n.t('screens.nations.toolbar.join')}
          disable={NationDetailsScreen.disableJoinButton(nation)}
          onPress={this.props.joinNation}
        />
        <NationActionButton
          iconSource={AssetsImage.Actions.leave}
          title={i18n.t('screens.nations.toolbar.leave')}
          disable={NationDetailsScreen.disableLeaveButton(nation)}
          onPress={this.props.leaveNation}
        />
      </View>
    );
  }
  render() {
    const nation = openedNation(this.props);

    if (!nation) {
      this.props.navigator.pop();
      return <BackgroundImage />;
    }

    const status = resolveStatus(nation);
    const statusDescription = (status !== null ? i18n.ifExists(`screens.nationDetails.statusDescription.${status.key}`) : '');

    return (
      <View style={styles.screenContainer}>
        <BackgroundImage />
        <FakeNavigationBar />
        <View style={styles.bodyContainer}>
          <ScreenTitle title={nation.nationName} />

          <ScrollView>
            {statusDescription !== '' && NationDetailsScreen.buildStatusPanel(statusDescription)}

            {NationDetailsScreen.buildAboutView(nation)}
            {NationDetailsScreen.buildCitizenPanel(nation)}
            {NationDetailsScreen.buildGovernmentalStructureView(nation)}
            {NationDetailsScreen.buildFactsView(nation)}
          </ScrollView>
        </View>
        {this.buildTabBar()}
      </View>
    );
  }
}

NationDetailsScreen.defaultProps = {
  joinNation: () => null,
  leaveNation: () => null,
  deleteDraft: () => null,
  submitDraft: () => null,
  openNationChat: () => null,
};

export default NationDetailsScreen;
