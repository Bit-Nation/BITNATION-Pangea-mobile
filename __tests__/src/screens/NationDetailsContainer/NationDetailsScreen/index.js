import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import NationDetailsScreen from '../../../../../src/screens/NationDetailsContainer/NationDetailsScreen';

describe('DetailNationScreen', () => {
  let detailsScreen;

  const initialStateMock = {
    nations: {
      nations: [{
        _id: 1,
        idInSmartContract: 1,
      }],
      openedNationId: 1,
    },
  };
  const storeMock = configureStore([]);
  const propsMock = {
    navigator: {
      pop: jest.fn(),
      setButtons: jest.fn(),
    },
    isDraft: false,
    joinNation: jest.fn(),
    leaveNation: jest.fn(),
    deleteDraft: jest.fn(),
    submitDraft: jest.fn(),
    openNationChat: jest.fn(),
  };

  beforeEach(() => {
    detailsScreen = shallow(<NationDetailsScreen
      {...propsMock}
      store={storeMock(initialStateMock)}
    />);
  });

  test('Rendering', () => {
    expect(detailsScreen).toMatchSnapshot();
  });
});
