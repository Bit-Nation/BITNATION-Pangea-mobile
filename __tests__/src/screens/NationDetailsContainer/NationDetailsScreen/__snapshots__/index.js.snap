// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailNationScreen Rendering 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NationDetailsScreen
    deleteDraft={[MockFunction]}
    isDraft={false}
    joinNation={[MockFunction]}
    leaveNation={[MockFunction]}
    navigator={
      Object {
        "pop": [MockFunction] {
          "calls": Array [
            Array [],
          ],
        },
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      }
    }
    openNationChat={[MockFunction]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    submitDraft={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "source": Object {
        "testUri": "../../../src/assets/images/background.jpg",
      },
      "style": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "source": Object {
          "testUri": "../../../src/assets/images/background.jpg",
        },
        "style": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
